### ISSUE TITLE
Unexpected deprecation warning after upgrading to TypeScript v4.2.2

### ISSUE THREAD

**Issue by Julien-Marcou** [microsoft/TypeScript#43053]
# Bug Report

### 🔎 Search Terms

RxJS Deprecation Warning TSLint ESLint VSCode Signature Doc Tags DocTags

### 🕗 Version & Regression Information

Since TypeScript v4.2.2, I get deprecation warnings with RxJS, that I was not getting in TypeScript v4.1.5

I first opened an issue on RxJS ([here](https://github.com/ReactiveX/rxjs/issues/6060)), but it may be a TypeScript issue

### 💻 Code

```ts
import { Observable, Subject } from 'rxjs';

const somethingHappened: Subject<void> = new Subject();
const somethingHappened$: Observable<void> = somethingHappened.asObservable();

// Deprecation warning on the next line
somethingHappened$.subscribe(() => {
  console.log('something happened');
});

somethingHappened.next();
```

### 🙁 Actual behavior

When I subscribe to the observable, I get the deprecation warning `subscribe is deprecated: Use an observer instead of a complete callback` even though I'm not using any complete callback

It occurs with both TSLint v6.1.3 and ESLint v7.21.0 (using eslint-plugin-deprecation v1.2.0)

Here are the signatures for the `subscribe` method in RxJS v6.6.6 :

```ts
subscribe(observer?: PartialObserver<T>): Subscription;
/** @deprecated Use an observer instead of a complete callback */
subscribe(next: null | undefined, error: null | undefined, complete: () => void): Subscription;
/** @deprecated Use an observer instead of an error callback */
subscribe(next: null | undefined, error: (error: any) => void, complete?: () => void): Subscription;
/** @deprecated Use an observer instead of a complete callback */
subscribe(next: (value: T) => void, error: null | undefined, complete: () => void): Subscription;
subscribe(next?: (value: T) => void, error?: (error: any) => void, complete?: () => void): Subscription;
```

If I understand this correctly, in my case it should be using this signature :

```ts
subscribe(next?: (value: T) => void, error?: (error: any) => void, complete?: () => void): Subscription;
```

Which is not deprecated

### 🙂 Expected behavior

There should be no deprecation warning

**Comment by gspat24** [comment]
+1

**Comment by nzbin** [comment]
+1

**Comment by itayperry** [comment]
Thank you for opening this issue - I have a huge Angular project with endless `tslint` deprecation warnings 🚀

**Comment by idamachmadfaizin** [comment]
> I have found this issue with upgrade vscode.

Yeah, I think since version: 1.54.1

**Comment by s4r14k** [comment]
Same here after upgrade vscode

**Comment by ak274** [comment]
After updating VSCode to 1.54.1. I started seeing these deprecation warnings.

**Comment by amosISA** [comment]
Same here.

**Comment by cvarona** [comment]
Same here

**Comment by zulihan** [comment]
Hi !
Same here. 
Visual studio Code:
Version: 1.54.0
Commit: 42e27fe5cdc58539dad9867970326a297eb8cacf
"tslint": "~6.1.0"

**Comment by Jonah-Jordan** [comment]
Same here

**Comment by adamkoncz** [comment]
Same issue after VSCode update

Version: 1.54.1 (user setup)
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:38:31.419Z

**Comment by jrdutton** [comment]
Ditto

Version: 1.54.1 (user setup)
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:38:31.419Z

**Comment by MikaStark** [comment]
Same as @jrdutton 

Version: 1.54.1 (user setup)
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:38:31.419Z

**Comment by Julien-Marcou** [comment]
> VS Code now ships with TypeScript 4.2.2. This major update includes many TypeScript language improvements, along with many improvements and bug fixes for JavaScript and TypeScript tooling.

FYI: The issue is still present with TypeScript 4.2.3

**Comment by ebastuart** [comment]
Same issue here: 
no matter if I use 
`.subscribe((value) => { /* code */ });`
or
`.subscribe({next: (value) => { /* code */ }}):`

looking forward for a fix. thank

**Comment by orta** [comment]
Hi folks, I'm not _100%_ that this is a TypeScript issue but an issue in both tslint and eslint which maybe aren't checking whether the _used_ version of a function is deprecated. 

When you make a blank project with:

```sh
mkdir 43053
yarn init -y
yarn add typescript rxjs
touch index.ts
code . 
``` 

Then open the file:

![Screen Shot 2021-03-09 at 3 09 34 PM](https://user-images.githubusercontent.com/49038/110491735-77c3fb80-80e9-11eb-9d97-ab00fff5e022.png)

We don't report to vscode that the former example is deprecated, which is what would happen if TypeScript thought it was deprecated.

-- 

That said, there's a reasonable possibility that the underlaying deprecations API usage changed in 4.2 though which might have caused this, which I'll be looking into next

**Comment by wnabil** [comment]
Thank god I thought I need to update the subscriptions to a new syntax ...

**Comment by orta** [comment]
So far, I've narrowed it down to being a change in the returned values from `getJsDocTags` and `getResolvedSignature.

With this script:

```ts
import * as ts from "typescript";

// Loop through the AST to find call expressions:
function traverse(sourceFile: ts.SourceFile) {
  lookAtNode(sourceFile);

  function lookAtNode(node: ts.Node) {
    if (ts.isCallExpression(node)) {
      check(node)
    }

    ts.forEachChild(node, lookAtNode);
  }
}

// Setup TypeScript environment
let program = ts.createProgram(["index.ts"], {});
let checker = program.getTypeChecker();
const sourceFile = program.getSourceFile("index.ts")!
traverse(sourceFile);

// Look at call expressions to see if they're deprecated based on code in tslint and SonarJS
function check(callExpression: ts.CallExpression) {
  const tc = checker;
  console.log("\n\nLooking at: ", callExpression.getText())

  const signature = tc.getResolvedSignature(callExpression);
  if (signature) {
    const tags = signature.getJsDocTags()
    const deprecated = tags.find(t => t.name === "deprecated")

    if (deprecated) {
      console.log("Deprecation from `getJsDocTags`", deprecated)
    }
  }
}
```

In 4.1:

```
Looking at:  somethingHappened.asObservable()


Looking at:  somethingHappened$.subscribe(() => {
  console.log('something happened');
})


Looking at:  console.log('something happened')


Looking at:  somethingHappened$.subscribe(null, undefined, () => {})
Deprecation from `getJsDocTags` {
  name: 'deprecated',
  text: 'Use an observer instead of a complete callback'
}
```

In 4.2:

```

Looking at:  somethingHappened.asObservable()


Looking at:  somethingHappened$.subscribe(() => {
  console.log('something happened');
})
Deprecation from `getJsDocTags` {
  name: 'deprecated',
  text: 'Use an observer instead of a complete callback'
}


Looking at:  console.log('something happened')


Looking at:  somethingHappened$.subscribe(null, undefined, () => {})
Deprecation from `getJsDocTags` {
  name: 'deprecated',
  text: 'Use an observer instead of a complete callback'
}


Looking at:  somethingHappened.next()
```

It's _unlikely_ this output is what we want, so hold off on new issues for other projects.

**Comment by Julien-Marcou** [comment]
Yes, I was about to say, even though VSCode no longer report a deprecation warning after disabling TSLint & ESLint, changing between TypeScript 4.1.5 & 4.2.2 changes the content of the "documention popup" when hovering the `subscribe` method.

Documention popup with TypeScript 4.1.5 :
![image](https://user-images.githubusercontent.com/4369075/110501177-34be5400-80fa-11eb-8c92-65ce389ffe16.png)

Documention popup with TypeScript 4.2.2 :
![image](https://user-images.githubusercontent.com/4369075/110501467-7fd86700-80fa-11eb-91c6-01d2697b4889.png)

Also, ESLint has no dependency on TypeScript, it is then more likely to be related to the [eslint-plugin-deprecation](https://github.com/gund/eslint-plugin-deprecation) which has been made based on [SonarJS](https://github.com/SonarSource/SonarJS) (in addition I just saw that an issue has been opened about [TypeScript 4.2 support for SonarJS](https://github.com/SonarSource/SonarJS/issues/2511) which may be related to the deprecation warnings we are seeing)

**Comment by orta** [comment]
OK, this looks like is an un-expected side-effect of https://github.com/microsoft/TypeScript/pull/42098

Repro'd without the whole rxjs pipeline:

```
interface PartialObserver<T> {}
interface Subscription {}
interface Unsubscribable {}

export interface Subscribable<T> {
  subscribe(observer?: PartialObserver<T>): Unsubscribable;
  /** @deprecated 1 Use an observer instead of a complete callback */
  subscribe(next: null | undefined, error: null | undefined, complete: () => void): Unsubscribable;
  /** @deprecated 2 Use an observer instead of an error callback */
  subscribe(next: null | undefined, error: (error: any) => void, complete?: () => void): Unsubscribable;
  /** @deprecated 3 Use an observer instead of a complete callback */
  subscribe(next: (value: T) => void, error: null | undefined, complete: () => void): Unsubscribable;
  subscribe(next?: (value: T) => void, error?: (error: any) => void, complete?: () => void): Unsubscribable;
}

interface ThingWithDeprecations<T> extends Subscribable<T> {
    subscribe(observer?: PartialObserver<T>): Subscription;
    /** @deprecated Use an observer instead of a complete callback */
    subscribe(next: null | undefined, error: null | undefined, complete: () => void): Subscription;
    /** @deprecated Use an observer instead of an error callback */
    subscribe(next: null | undefined, error: (error: any) => void, complete?: () => void): Subscription;
    /** @deprecated Use an observer instead of a complete callback */
    subscribe(next: (value: T) => void, error: null | undefined, complete: () => void): Subscription;
    subscribe(next?: (value: T) => void, error?: (error: any) => void, complete?: () => void): Subscription;
}

declare const a: ThingWithDeprecations<void>

// Fails with the above script
a.subscribe(() => {
  console.log('something happened');
});
```

**Comment by ubugnu** [comment]
Same here, happened after updating to vscode 1.54.1

**Comment by mhamri** [comment]
+1

**Comment by mhamri** [comment]
@sandersn can we open this issue? It come back again.

**Comment by Julien-Marcou** [comment]
Actually, it never disappeared ^^

The fix has not been merged for the 4.2.X branch yet, it has only been merged for the next 4.3.X release.

You can track the merge progress for the 4.2.X branch here : https://github.com/microsoft/TypeScript/pull/43180

**Comment by mhamri** [comment]
so why this issue is closed?

**Comment by anuragarwalkar** [comment]
facing same issue

**Comment by VincenzoLomba** [comment]
Same here, problem not solved yet.

**Comment by moscoso** [comment]
Facing same issue for  Typescript 4.2.3

Vs Code versions:
Version: 1.54.2 (user setup)
Commit: fd6f3bce6709b121a895d042d343d71f317d74e7
Date: 2021-03-11T00:56:19.848Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.18363

**Comment by DrakeAnglin** [comment]
+1

**Comment by Manu1909** [comment]
+1

**Comment by 1antares1** [comment]
Same here, problem not solved yet (waiting for 4.3.X release).

**Comment by adelynx** [comment]
Same issue here I have tried everything no chance.

**Comment by benjaminao** [comment]
Same issue here! Please let us know how avoid this warning!

**Comment by Julien-Marcou** [comment]
If you really need TypeScript 4.2 -> Wait for a patch

If you don't need TypeScript 4.2 -> You can simply downgrade your project and VSCode to use TypeScript 4.1

- Open a TypeScript file
- Click on the current version of TypeScript you are using (bottom right corner of VSCode)
![](https://user-images.githubusercontent.com/4369075/112892563-c1e14100-90d9-11eb-9d02-d3d65303090a.png)
- Choose "Select TypeScript Version..."
![](https://user-images.githubusercontent.com/4369075/112893593-2b158400-90db-11eb-844a-93ae91977699.png)
- Choose "Use Workspace Version"
![](https://user-images.githubusercontent.com/4369075/112892974-5481e000-90da-11eb-81e4-eddc869d1660.png)

**Comment by orta** [comment]
You can also use the nightly version of the TS extension: https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-next

**Comment by fromage9747** [comment]
Updated today to version 1.55 of VS code.  The issue still persists.

**Comment by Julien-Marcou** [comment]
TypeScript 4.2.4 has now been released, so we can close this.

A VSCode update should follow in about a week.

**Comment by MartinHalata** [comment]
I did upgrade to typescript 4.2.4, but problem still persist
![image](https://user-images.githubusercontent.com/79250857/114144728-edec9580-9915-11eb-9c0f-05d2ff1e8002.png)

**Comment by BruneXX** [comment]
Hi guys, I've recently installed VSCode 1.55.2 and typescript v4.2.4 but I'm still facing this issue, anyone have any clue about this? will be appreaciated. Thanks

**Comment by mirzasetiyono** [comment]
have this issue solved?

**Comment by radiumrasheed** [comment]
can't seem to figure how to set the right version for sonar.
what's the workaround for this, please?

**Comment by Julien-Marcou** [comment]
Using this https://github.com/microsoft/TypeScript/issues/43053#issuecomment-809677495 to force VScode to use TypeScript v4.2.4 instead of v4.2.3 I can confirm the issue has been fixed.

Unfortunately the branch for VSCode v1.55 still hasn't been updated with TypeScript v4.2.4.
I opened a request on the VSCode repository https://github.com/microsoft/vscode/issues/121583 to get this fixed in VSCode 1.55.3 hopefully.

**Comment by my-lalex** [comment]
My 2 cents, I'm not sure this is the right place to post this, but a deprecation warning still occurs on a specific situation.
Actually, it's in the 'body-parser' types of the Express plugin
```typescript
/** @deprecated */
declare function bodyParser(
    options?: bodyParser.OptionsJson & bodyParser.OptionsText & bodyParser.OptionsUrlencoded,
): NextHandleFunction;

declare namespace bodyParser {
    //...
    /**
     * Returns middleware that only parses json and only looks at requests
     * where the Content-Type header matches the type option.
     */
    function json(options?: OptionsJson): NextHandleFunction;
    //...
}
```

Using `bodyParser.json()`in VSCode display the `bodyParser`part as stroked, and the `json()` part as it should (unstroked)...
The deprecation feature, in this case, doesn't understand that the`bodyParser`refer to the namespace and not the function...

**Comment by fromage9747** [comment]
Today's update to VSCode sees it running Typescript 4.2.4 and the deprecation warning got Subscribe is no longer present!

**Comment by SpeedoPasanen** [comment]
This is still happening at least with the `tap` function from `rxjs`. VSCode 1.58.2. TS version in project: 4.3.5

Deprecation warning shows up regardless of which overloaded signature you use. Should not show up for the last two.

```
interval(5000)
  .pipe(
     tap(console.log), // "deprecated - Use an observer instead of a complete callback"
     tap({ next: console.log, error: console.log, complete: console.log }) // "deprecated
  )
  .subscribe(() => console.log('')); // this shows no warnings anymore
```

`rxjs/tap.d.ts`:
```
import { MonoTypeOperatorFunction, PartialObserver } from '../types';
/** @deprecated Use an observer instead of a complete callback */
export declare function tap<T>(next: null | undefined, error: null | undefined, complete: () => void): MonoTypeOperatorFunction<T>;
/** @deprecated Use an observer instead of an error callback */
export declare function tap<T>(next: null | undefined, error: (error: any) => void, complete?: () => void): MonoTypeOperatorFunction<T>;
/** @deprecated Use an observer instead of a complete callback */
export declare function tap<T>(next: (value: T) => void, error: null | undefined, complete: () => void): MonoTypeOperatorFunction<T>;
export declare function tap<T>(next?: (x: T) => void, error?: (e: any) => void, complete?: () => void): MonoTypeOperatorFunction<T>;
export declare function tap<T>(observer: PartialObserver<T>): MonoTypeOperatorFunction<T>;
```

**Comment by antoniobuttiglione** [comment]
Plus one here, I'm facing the same issue. The problem is that on RxJS website is not explained well how the new code should look like.
